Adds examples using <script> tags in CAPA Checkbox and Multiple Choice problems

pomegranited opened the pull request:
## [OSPR-1468](https://openedx.atlassian.net/browse/OSPR-1468)

Describes and demonstrates how to use `<script>` tags in Checkbox and Multiple Choice problems.

cf. [edx-platform PR #13534](https://github.com/edx/edx-platform/pull/13534)

### Date Needed

We're aiming to get [edx-platform PR #13534](https://github.com/edx/edx-platform/pull/13534) merged before the 17 Oct 2016 release branch is cut.

### Reviewers

Possible roles follow. The PR submitter checks the boxes after each reviewer finishes and gives :+1:. 

- [x] OpenCraft: @smarnach
- [x] Doc team review (sanity check, copy edit): @lamagnifica , @srpearce 
- [x] Product review: @sstack22 

CC @gsong 

### Testing

- [x] Ran ./run_tests.sh without warnings or errors

### Post-review

- [ ] Add a comment with the description of this change or link this PR to the next release notes task.
- [ ] Squash commits

gsong commented:
Thanks for the pull request, @pomegranited! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-documentation&number=1282

pomegranited commented:
@gsong I intended to include this change in the existing [OSPR-1468](https://openedx.atlassian.net/browse/OSPR-1468).  Is that ok?

gsong commented:
@pomegranited That works.

smarnach commented on en_us/shared/exercises_tools/checkbox.rst:
```diff
@@ -937,3 +937,61 @@ Children
+``]]>`` markers, to indicate that the enclosed code should not be interpreted
+as XML.
+
+For example, the following OLX uses a python script to determine the choices
```
This makes it sound like it's possible to use other languages than Python as well, but I don't think that's true.  I'd suggest to state explicitly that only Python is supported.

(Also, Python should be capitalised.)

pomegranited replied:
Addressed with e145e86.

smarnach commented on en_us/shared/exercises_tools/checkbox.rst:
```diff
@@ -937,3 +937,61 @@ Children
+Using the Script Tag
+====================
+
+You can use the `<script>` tag to set attributes and options for your checkbox
```
Should use double back ticks around `<script>`.  This is reStructured Text, not Markdown.

lamagnifica replied:
is it an ^HTML script tag? we tried to standardize so that we use "tag" only for HTML, but use "element" for OLX/XML

pomegranited replied:
@smarnach @lamagnifica Addressed with 8f99e3a.

lamagnifica commented on en_us/shared/exercises_tools/checkbox.rst:
```diff
@@ -937,3 +937,61 @@ Children
+**************************************
+Advanced Options for Checkbox Problems
+**************************************
+
```
Possibly up around line 700 add a cross reference to this info? 
like,
.. note:: You can also set attributes and options for a checkbox problem by adding a ``<script>`` tag. For more information, see :ref:`whatever you name the anchor for this section`.

pomegranited replied:
Addressed with ecab0f8.

lamagnifica commented on en_us/shared/exercises_tools/checkbox.rst:
```diff
@@ -937,3 +937,61 @@ Children
+
+You can use the `<script>` tag to set attributes and options for your checkbox
+problem.  The code in the ``<script>`` tag is run on the server before the
+problem is shown to learners.
```
why would you choose to use ``<script>`` instead of just setting the attributes and options yourself in the OLX?  An example or two might be helpful here.

pomegranited replied:
Addressed with e145e86.

lamagnifica commented on en_us/shared/exercises_tools/checkbox.rst:
```diff
@@ -937,3 +937,61 @@ Children
+        text3 = "None of the above statements is true."
+        ]]>
+        </script>
+        <p>Which statements about the number $a+$b are true? Select all that apply.</p>
```
change < p> and < /p> to < label> and < /label>
< label> element is now I believe a child of choiceresponse

pomegranited replied:
That's cool..  probably more accessible this way?  Addressed with 3c2714d.

lamagnifica replied:
exactly!

lamagnifica commented on en_us/shared/exercises_tools/checkbox.rst:
```diff
@@ -937,3 +937,61 @@ Children
+            <choice correct="$ok3">$text3 ... (should be $ok3)</choice>
+          </checkboxgroup>
+        </choiceresponse>
+        <p><strong>Comment:</strong> In the <code>choice</code> tags for this problem, both the <code>correct</code> attribute and tag content are set set using Python.</p>
```
would this text be shown to learners in the problem component? if so, and this information is important for understanding how to approach answering the problem, it should be in the <description> element immediately after the < label>

also, strike second "set"

pomegranited replied:
This text is not needed, so I incorporated it into the "why to use the script tag" explanation for e145e86.

lamagnifica commented on en_us/shared/exercises_tools/multiple_choice.rst:
```diff
@@ -984,3 +984,50 @@ explanation ID.
+
+You can use the ``<script>`` tag to set attributes and options for your multiple
+choice problem.  The code in the ``<script>`` tag is run on the server before
+the problem is shown to learners.
```
apply all similar suggestions that make sense, including the cross reference, examples or background on the value of using this feature, etc. throughout this file

pomegranited replied:
@lamagnifica Yep, I made changes to both files as I went along.

lamagnifica commented on en_us/shared/exercises_tools/checkbox.rst:
```diff
@@ -937,3 +937,61 @@ Children
+Use the Advanced Editor to Configure the Script Tag
+***************************************************
+
+You must use the :ref:`advanced editor<Advanced Editor>` to use the
```
change "use the" to "configure a" (?) 
(makes this sentence parallel to the heading, and to avoid using the same verb 2x in one sentence)

lamagnifica commented on en_us/shared/exercises_tools/multiple_choice.rst:
```diff
@@ -984,3 +984,50 @@ explanation ID.
+as XML.
+
+For example, the following OLX uses a python script to determine the choices
+available for the problem, and sets their "correctness":
```
change : to .
(it's just an edX style choice we made, to use complete sentences)

pomegranited replied:
Addressed with e145e86.

lamagnifica commented on en_us/shared/exercises_tools/checkbox.rst:
```diff
@@ -937,3 +937,61 @@ Children
+as XML.
+
+For example, the following OLX uses a python script to determine the choices
+available for the problem, and sets their "correctness":
```
change : to .

pomegranited replied:
Addressed with e145e86.

lamagnifica commented on en_us/shared/exercises_tools/multiple_choice.rst:
```diff
@@ -984,3 +984,50 @@ explanation ID.
+        text1 = "$a + $b is odd"
+        ]]>
+        </script>
+        <p>Is $a+$b even or odd?</p>
```
change p to label, also I believe label goes inside the response element

lamagnifica commented on en_us/shared/exercises_tools/multiple_choice.rst:
```diff
@@ -984,3 +984,50 @@ explanation ID.
+            <choice correct="$ok1">$text1 ... (should be $ok1)</choice>
+          </choicegroup>
+        </multiplechoiceresponse>
+        <p><strong>Comment:</strong> In the <code>choice</code> tags for this problem, both the <code>correct</code> attribute and tag content are set set using Python.</p>
```
strike second "set", move to a description if that seems like the right thing to do

pomegranited replied:
This text is not needed, so I incorporated it into the "why to use the script tag" explanation for e145e86.

lamagnifica commented:
@marcotuts @sstack22 here's the companion doc PR

sstack22 commented:
@lamagnifica - do we generally specify support levels for XML authoring options?

lamagnifica commented:
@sstack22 we do not, as our strong preference is to scope docs to fully supported features only whenever possible. That said, there are many problems in older/frequently rerun courses that use no-longer-maintained and not-fully-supported OLX. Here's an example of how we document but do not document something: http://edx.readthedocs.io/projects/edx-partner-course-staff/en/latest/exercises_tools/numerical_input.html#formulaequationinput

pomegranited commented:
@smarnach @lamagnifica I believe I've addressed all of your comments.  Thank you for your reviews!

pomegranited commented:
@sstack22 I'm not clear on whether [your comment](https://github.com/edx/edx-platform/pull/13534#issuecomment-252041782) means that I should move this content into the OLX guide only, or whether it's ok to leave it as-is?

@marcotuts @sstack22 @srpearce Did you need any further changes to these docs?

sstack22 commented:
@pomegranited  - I think it's ok to leave as in, 👍  from me!

lamagnifica commented on en_us/shared/exercises_tools/checkbox.rst:
```diff
@@ -700,6 +700,9 @@ halves.
 Checkbox Problem OLX Reference
 ******************************
 
+.. note:: You can also set attributes and options by adding a ``<script>`` tag.
```
element?

pomegranited replied:
Gah, sorry about that!  Fixed with [f3c361e L703]https://github.com/edx/edx-documentation/pull/1282/commits/f3c361e3797d8f4c44b7284bc23b65614010d9dd#diff-ef222f38fb9bcc1257536ca485a816d0R703).

lamagnifica commented on en_us/shared/exercises_tools/checkbox.rst:
```diff
@@ -937,3 +940,67 @@ Children
+        </script>
+        <choiceresponse>
+          <label>Which statements about the number $a+$b are true? Select all that apply.</label>
+          <checkboxgroup label="Mark the true statements" direction="vertical">
```
we no longer use the label attribute (the label element replaced it)

pomegranited replied:
Oo awesome.. the XML is looking very tidy now!  Fixed with [f3c361e L999](https://github.com/edx/edx-documentation/pull/1282/commits/f3c361e3797d8f4c44b7284bc23b65614010d9dd#diff-ef222f38fb9bcc1257536ca485a816d0R999).

lamagnifica replied:
😄 lots of a11y work this past summer!

lamagnifica commented on en_us/shared/exercises_tools/multiple_choice.rst:
```diff
@@ -456,6 +456,9 @@ provides partial credit of 25% for an answer option.
 Multiple Choice Problem OLX Reference
 *************************************
 
+.. note:: You can also set attributes and options by adding a ``<script>`` tag.
```
element

pomegranited replied:
Fixed with [f3c361e L459](https://github.com/edx/edx-documentation/pull/1282/commits/f3c361e3797d8f4c44b7284bc23b65614010d9dd#diff-2c65bfb059143d63aad49b5ed0dbd4a0R459).

lamagnifica commented on en_us/shared/exercises_tools/multiple_choice.rst:
```diff
@@ -984,3 +987,56 @@ explanation ID.
+as XML.
+
+The code in the ``<script>`` element is run on the server before the problem is
+shown to learners.  Note that only Pythons script types are supported.
```
Python singular

pomegranited replied:
Eesh.. Thank you for spotting that!  Fixed with [f3c361e L1013](https://github.com/edx/edx-documentation/pull/1282/commits/f3c361e3797d8f4c44b7284bc23b65614010d9dd#diff-2c65bfb059143d63aad49b5ed0dbd4a0R1013).

lamagnifica commented on en_us/shared/exercises_tools/multiple_choice.rst:
```diff
@@ -984,3 +987,56 @@ explanation ID.
+        </script>
+        <multiplechoiceresponse>
+          <label>Is $a+$b even or odd?</label>
+          <choicegroup label="Select the true statement" type="MultipleChoice">
```
no need for the label attribute

pomegranited replied:
Fixed with [f3c361e L1037](https://github.com/edx/edx-documentation/pull/1282/commits/f3c361e3797d8f4c44b7284bc23b65614010d9dd#diff-2c65bfb059143d63aad49b5ed0dbd4a0R1037).

pomegranited commented:
@lamagnifica Thank you for catching those typos and issues!  I believe I've addressed your comments, and have fixed the failing tests.

Let me know when it's ok to squash these commits?  And please remind me, do you handle the merge once the [feature PR](https://github.com/edx/edx-platform/pull/13534#issuecomment-250472862) is done?

srpearce commented:
@pomegranited Nothing else from me. Thanks!

pomegranited commented:
Thanks, @srpearce !

lamagnifica commented:
looks terrific, thanks @pomegranited ! 👍 Please let me know when you merge, or that I should go ahead and merge after you squash. That will be my cue to build the 5 (!) affected guides.

pomegranited commented:
Thanks, @lamagnifica !  I made one further change (a938681) due to an issue raised by @staubina.  See conversation starting with https://github.com/edx/edx-platform/pull/13534#issuecomment-253301785.

Will squash once you're both happy with this.

lamagnifica commented:
ok by me!

pomegranited commented:
Thanks, @cahrens !  I've squashed the commits in preparation for merge.

pomegranited commented:
@cahrens Merged the feature PR https://github.com/edx/edx-platform/pull/13534, so this is ready for the next release cut.

lamagnifica commented:
@pomegranited one last request. I crafted this blurb for the release notes next week. Please let me know if there is anything you'd like to add/change/correct/delete etc. 
"This release includes an enhancement to the core checkbox and multiple choice problem types. When you edit these problems in the advanced editor, you can now set attributes and options by adding a ``<script>`` element. For more information, see :ref:`partners:Using the Script Element in Checkbox Problems` or :ref:`partners:Using the Script Element in Multiple Choice Problems`."

pomegranited commented:
@lamagnifica Your blurb sounds good to me. Technically, the `<script>` functionality has been there all along; we just fixed a bug with Checkbox problems and the `correct` attribute.

But if an undocumented tree falls in a forest, did it really exist?  😄

catong commented:
@lamagnifica This PR will merge for the Oct 18, yes?

lamagnifica commented:
@catong yes, that is the intention -- I believe I added a comment in the release notes story to make sure this got did. Do you want me to merge it now?

edx-webhook commented:
Thanks for the pull request, @pomegranited! I've created [OSPR-5328](https://openedx.atlassian.net/browse/OSPR-5328) to keep track of it in JIRA, where we prioritize reviews. Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

nedbat commented:
Hi, sorry for the new comment on this old pull request. We're testing out re-scanning contributions, and our bot added a confusing message that you can ignore. We'll fix that. Sorry about the noise.
